### ISSUE TITLE
Adopt the StatusNotifierItem specification

### ISSUE THREAD

**Issue by Airblader** [i3/i3#2088]
The XEmbed based system tray specification is horrible. We've noticed this when trying to allow transparency, and so has [awesome](https://github.com/awesomeWM/awesome/issues/363).

Even all the big players, namely Unity, KDE and Gnome, started deprecating the tray specification and seem to be betting on the replacement instead: the [StatusNotifierItem specification](http://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/).

I haven't looked at the spec yet, but we should consider starting to implement this as well.

References
[1] http://blog.martin-graesslin.com/blog/2014/03/system-tray-in-plasma-next/
[2] https://bugzilla.gnome.org/show_bug.cgi?id=734826
[3] https://mail.gnome.org/archives/gnome-flashback-list/2015-April/msg00009.html

**Comment by stapelberg** [comment]
Thanks for bringing this up, I think implementing this sounds reasonable in theory.

Given that the StatusNotifierItem spec seems to be based on dbus, we’ll need to have a careful look at which dbus libraries are out there that we can depend on.

**Comment by Airblader** [comment]
Alright. I'm not going to work on this right away, too much on my plate already, but I'll mark it as accepted for now under the usual restrictions and disclaimers.

**Comment by Airblader** [comment]
@stapelberg We already depend on glib which comes with a dbus API, so I think we should just use that one.

**Comment by stapelberg** [comment]
I haven’t used that API, but the fact that we already depend on glib is a strong plus of course. I guess we should try using the glib dbus API.

**Comment by psychon** [comment]
GDbus is actually part of Gio and not GLib (but the two are clearly related).

**Comment by reitzig** [comment]
Got here after searching for issues related to the problem I'm seeing with transparent icons:

![screenshot](https://user-images.githubusercontent.com/4246780/42320980-8591444c-8057-11e8-86f0-8f0112010871.png)

This keeps popping up: #2325, #3197 It _is_ an awful user experience.

Is there any timeline on this issue, or should we open one to handle transparent icons in _some_ way? (Just draw a square of background color behind each icon...)

**Comment by psychon** [comment]
@reitzig You are only seeing this with Qt-using applications, right? If so, I have good news for you: I remember seeing some patch that fixes this in Qt. A quick search found: https://codereview.qt-project.org/#/c/139051/ (which seems to be stalled, sigh).

**Comment by reitzig** [comment]
@psychon I don't have enough data to say so. The one app I notice this with is a Qt one, yes. Most other apps don't have a lot of transparency in their icons, though.

**Comment by BonaBeavis** [comment]
Bumping this, got the missing redraw with qt icons and the background of gtk3 icons is transparent(It should not be transparent).

**Comment by heroin-moose** [comment]
@Airblader, is this something you are considering to implement?

**Comment by Airblader** [comment]
Personally, no.

**Comment by Gigadoc2** [comment]
Maybe a bit off-topic, but since GNOME has [dropped status items as a whole](https://blogs.gnome.org/aday/2017/08/31/status-icons-and-gnome/) quite some time ago, I don't think this will ever get implemented.
Some of the reasons they named also apply to tiling wm users. In particular, you can't really control which icons you get (not every application lets you disable tray icons), and you can't control the order of these icons either.

Personally, I battled with the xembed problems as well for years, even went so far as to switch to sway + waybar primarily for the SNI icons, but even there they aren't all that nice.
I have finally settled with just keeping the windows of applications that would have status icons open on specific workspaces, turning off status icons entirely. So far, I have not encountered any application that *really* need status icons.

**Comment by crocket** [comment]
@Gigadoc2 What about redshift-gtk, ibus, fcitx, and parcellite?

**Comment by heroin-moose** [comment]
FYI, I cannot reproduce this on polybar.

**Comment by Gigadoc2** [comment]
> @Gigadoc2 What about redshift-gtk, ibus, fcitx, and parcellite? I don't yet know an alternative to status icon for displaying status or presence for those applications.

Personally, I just run redshift without any ui frontend, it works fine in the background. I've never used an input method, so I am not sure about those, but don't you usually activate them with a shortcut anyway? If so, you do you *really* need the icons still? For clipboard managers, I think there are some that can be used with just shortcuts as well (I found [clipmenu](https://github.com/cdown/clipmenu) with a quick search). Personally I don't use them at all anymore, due to security concerns (and wayland).

I mean, this is just personal preference and you may very validly have a different preference. I just wanted to encourage all those who were like past me to just try and live without status icons for a few weeks, and see if you still miss them.

It also depends on what you use the status icons for. If you use them for interacting with applications, I'd ask if that is really a convenient way to use applications, considering that the tray icons change order all the time? For tiling WMs setting up keybindings for apps comes to mind immediately.
If you use them to see whether your applications are running and/or working correctly, I'd ask whether you really want to see the status of all applications, or just a select few? In the latter case, you could use something like i3status' `run_watch` to watch for the existence of certain processes. This has the advantage that the status information will stay at the same place throughout reboots and the like.

Independent of all our opinions on status icons and the like, I also wanted to mention that GNOME has officially abandoned them, and this has probably an influence on the ecosystem outside of GNOME as well:
Most importantly, application developers close to the GNOME ecosystem will probably not implement status icons anymore, as they themselves can't even see them on their desktops.
This means, DE/WM developers outside of GNOME might also be more hesitant to work on status icon features, as a lot of apps will now not use them anymore.
KDE still has status icons though, so this might be dependent on how many "KDE-close" applications one uses.

**Comment by anowlcalledjosh** [comment]
Personally: being able to click icons is a desirable feature; I don't have the brainspace to memorise a keyboard shortcut for each of the indicator applets I use. I also don't want to have to write my own application to do something like display a list of WiFi networks, I have enough tiny custom scripts already.

Further, suggesting to use an i3status module (or similar) is a significant loss of functionality, since indicator applets can have colourful icons, which is significantly harder to do in i3status (even assuming that i3bar will properly display colour emoji, you still have to find a font which contains the characters you want to display).

It's also not reasonable to say that an i3bar-based solution is better because the icons will have a consistent order – there is no technical reason why i3bar cannot display tray icons in a deterministic order, it's just not implemented yet (there is a patch, but it's buggy, and I don't have the time or knowledge to debug it): #3573

In summary: yes, indicator applets suck, but all the alternatives I've seen proposed suck more. People still use indicator applets, so improving i3bar's handling of them seems like a sensible thing to do.

**Comment by reitzig** [comment]
2c: On my desktop PC, I could probably go without "notification area" icons (Flameshot, Redshit, Jetbrains Toolbox, Tresorit all require little interaction beyond opening the main windows --> Ctrl+D). On my work laptop, much more is going on: all of the above plus network manager applet, Bluetooth applet, Logitech Unifying applet -- some of which I frequently interact with.

Plus, if I ever were to create a UI interface to the numerous helper scripts or shell commands I have inadvertantly come to rely on with i3 -- say, a less crazy person wants to use my computer -- that would probably be the place to put a "start menu".

**Comment by kafran** [comment]
#### Question
Friends, how are you alternatively surviving without tray icons? 

#### Context
I use i3 on top of a Mate DE replacing Marco for i3 on a Linux Mint installation (this is the way I found to get things up and running fast). I'm thinking to change to Mint XFCE on the next release and replace Xfwm for i3, because Caja has been bugging me a lot to run on a Mate Session with i3 as WM.

#### The problem
Problem is that both Mate and XFCE has been gradually migrating to systray I think are not based on XEmbed. On XFCE its even worse as important services like volume control has migrated from daemon to applet. Even nm-applet has to be launched with --no-indicator to properly work on i3bar.

#### The future
So, it seems systray is something that will not survive the future. SwayBar doesn't have a good support for it. Gnome has get rid of it... So, whats the future of systray icons? And how are you adapting to it from a UX perspective. I use systray mostly for enable/disable bluetooth and to pair new devices, to look for wifi networks, to see if my dropbox is running and synchronizing w/o problems, to clipboard management, and to power-management. How can I manage this without having to bend my brain?

So, question is: how are you alternatively surviving without systray icons and how could I adapt to this not far away future?

**Comment by steils** [comment]
> Friends, how are you alternatively surviving without tray icons?

I use snixembed, https://git.sr.ht/~steef/snixembed

**Comment by stapelberg** [comment]
Okay, so to recap the current status:

* In 2015 it looked like this StatusNotifierItem is the future for trays. In the meantime, GNOME seems to have given up entirely on the concept of trays. What does adoption look like right now? Would we make possible any new use-case or better experience compared to today?

* The only thing currently speaking for adding support, IMO: Qt5 has a bug that results in rendering issues when using the older xembed protocol, so that might be worked around by implementing StatusNotifierItem.

* Given the complexity of e.g. https://github.com/swaywm/sway/pull/1234, and the fact that it seems like the only advantage is to work around a bug in Qt5, I’m tending towards not having this functionality in i3.

Is that accurate? Am I missing something?

**Comment by Airblader** [comment]
I still see a big demand for a tray among users and don't really understand or support Gnome's decision. Or, better put, I understand their motivation, but it misses some points which impact users, and users of window managers such as i3 even more so, IMO. 

Whether or not SNI still provides a significant improvement here — I'm not so sure anymore. It does trade quite a lot of complexity for seemingly minor improvements. As with X itself, it seems doubtful that XEmbed is going anywhere anytime soon.

**Comment by stapelberg** [comment]
I fully agree!

**Comment by keysmashes** [comment]
I was under the impression that SNI would allow for the same tray to be displayed on multiple outputs, which would be a significant improvement over the current situation for users with multiple monitors.

**Comment by nlgranger** [comment]
My 2ct contribution for this discussion:

- Gnome works to integrate apps so as to avoid the need for trays. i3 doesn't provide integration for volume control, network, battery status, etc. which can be implemented via tray icons, so they are useful IMHO.
- snixembed readme seems to imply it won't be developed anymore.
- Supporting both Xembed and StatusNotifierItem is extra work for app developers, but libappindicator can help for supported languages.
- Adding StatusNotifierItem would bring more feature parity between sway and i3.
- Supporting SNI seems like a lot of work indeed, is there anything that can be reused from the work done on sway?

**Comment by hyiltiz** [comment]
What are the technical difficulties in implementing support for SNI along with the currently used XEmbed based tray icons? I don't think looking up to GNOME or others are much meaningful.

**Comment by orestisfl** [comment]
Nothing necessarily "hard" but a lot of code needs to be written/ported and then debugged & maintained. We are currently a bit conservative with implementing new features that might lead to new bugs because we have a reduced amount of time.

However, I am sure that if somebody is motivated enough to develop & maintain a patch for SNI for i3, it would find a lot of usage (myself included).

**Comment by hyiltiz** [comment]
I think it is at least a skeptical idea to follow GNOME rather than a standard even though that standard came from freedesktop, which is closely related to GNOME. As far as I could find out, the system panels/bars/trays from KDE, XFCE, LXQT as well as third party ones such as polybar all support SNI. To be able to attract contribution, it would be helpful if we can describe and propose a way to implement it (specific details such as which functions in `xcb.c` needs a SNI cousin, and at what level a wrapper function dispatches into SNI or XEmbed protocols etc.) That could significantly lower the bar for anyone interested to give it a try over a few weekends. 

The following project, released under a permissive license, seems to provide a wrapper for SNI icons so they "act like" XEmbed style icons. In the worst case scenario, an i3 user compiles and runs this project (still trying; it didn't *simply* compile). A better way would be to maybe draw inspiration from these projects and some of the ones mentioned above so we could provide a minimal but ideally platform independent (not depending on gtk or qt libraries etc.) implementation.
https://git.sr.ht/~steef/snixembed

**Comment by psychon** [comment]
Since you mention snixembed, here is what happened when I took a look at it: https://github.com/awesomeWM/awesome/issues/2995#issuecomment-681979350

I also recommend reading the next post in that issue, because the snixembed author chimed in and wrote a bit about it.

The conclusion for me was: I cannot test my implementation, nothing follows the standard anyway, and the need for this protocol seems to have (partially) vanished. Thus, I stopped looking at it.

**Comment by Airblader** [comment]
Thanks @psychon. I have also long arrived at the conclusion that this is not worth implementing given the effort that has to go into it while considering the state of things.

Let's pull the plug on this.

**Comment by hyiltiz** [comment]
None of the KDE apps, including the one I care about such as Network Manager, KMix, Kgpg, blueman, and fcitx all now stopped showing their icons. I try not to use Electron apps, so I do not care much about them; I can imagine people saying the same to me about KDE apps, though.

**Comment by codeurimpulsif** [comment]
Hi, this method to show systray icons and menus are more and more used.

We speak about this for the last version (2.0.0) of Gajim [here](https://dev.gajim.org/gajim/gajim/-/issues/12197).

Would it be a good idea to reopen this issue to reconsider implementing it?

**Comment by stapelberg** [comment]
Looks like Gajim implemented StatusNotifierItem on their own (see `gajim/common/dbus/statusnotifieritem.py`) because GTK4 dropped support for XEmbed (but also does not offer support for StatusNotifierItem).

I’ll reopen this issue for discussion, but I don’t have any time to look into this situation more. 

I think a good start would be an analysis of which (popular) programs use which standard by now, and then we’d need somebody motivated and capable of contributing an implementation.

**Comment by pstray** [comment]
Dropbox is dropping XEMBED and will support just SNI starting from 25. April.

**Comment by kristijanhusak** [comment]
Here's a workaround for polybar users:
1. Install `dropbox-cli`
2. Add this script to your polybar scripts:
```sh
#!/bin/bash
status=$(dropbox-cli status)

if [[ "$status" == "Up to date" ]]; then
    echo "%{F#0000FF}%{F-}"
elif [[ "$status" == "Syncing"* ]]; then
    echo "%{F#0000FF}󰜷%{F-}"
elif [[ "$status" == "Connecting..." ]]; then
    echo "%{F#FFA500}󰇚%{F-}"
else
    echo "%{F#FF0000}%{F-}"
fi
```
3. Add a polybar module to your config:
```dosini
[module/dropbox]
type = custom/script
exec = ~/.config/polybar/dropbox.sh
```
4. Add `dropbox` to your `modules-right`

Here's how it looks for me:
1. Up to date - <img width="31" height="22" alt="Image" src="https://github.com/user-attachments/assets/2cc5aba1-b329-43b2-9bcf-9292a264ea3d" align="center" />

2. Syncing - <img width="38" height="28" alt="Image" src="https://github.com/user-attachments/assets/34773ac3-067d-4157-8b66-8b704b26233b" align="center"/>

3. Connecting - <img width="40" height="30" alt="Image" src="https://github.com/user-attachments/assets/7c1031ab-30da-4165-8841-11b678589ae5" align="center" />

4. Error/Disconnected - <img width="39" height="26" alt="Image" src="https://github.com/user-attachments/assets/3a9f90dc-1f77-4eaf-a7d5-94143e492bd3" align="center"/>

**Comment by deen1** [comment]
Via [this comment](https://github.com/i3/i3/discussions/6430#discussioncomment-13275920) in the discussion pages, installing [snixembed](https://git.sr.ht/~steef/snixembed) solves the Dropbox problem for me. (I also haven't had any issues with `dunst` as described there.)

**Comment by typhoon2099** [comment]
I'm so sure that `blueman-tray` used to appear in my systray up until the last few months. I still prefer to use systray for wifi/bluetooth management, but I guess I will have to either take a look at `snixembed` or try to replace my usual workflows with dmenu.
